File: openmole-ui-ide/openmole-ide-ant/org.openmole.ui.ide/src/org/openmole/ui/ide/workflow/provider/DnDProvider.java
/*
 *  Copyright (C) 2010 Mathieu Leclaire <mathieu.leclaire@openmole.fr>
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.openmole.ui.ide.workflow.provider;

import java.awt.Graphics2D;
import java.awt.Image;
import java.awt.Point;
import java.awt.Rectangle;
import java.awt.datatransfer.Transferable;
import java.awt.geom.AffineTransform;
import javax.swing.JComponent;
import org.netbeans.api.visual.action.AcceptProvider;
import org.netbeans.api.visual.action.ConnectorState;
import org.netbeans.api.visual.widget.Widget;
import org.openmole.ui.ide.workflow.implementation.MoleScene;
import org.openmole.ui.ide.workflow.model.IGenericTaskModelUI;
//import org.openide.awt.StatusDisplayer;

/**
 *
 * @author Mathieu Leclaire <mathieu.leclaire@openmole.fr>
 */
public abstract class DnDProvider implements AcceptProvider {
    protected MoleScene scene;

    public DnDProvider(MoleScene molescene) {
        this.scene = molescene;
    }

    @Override
    public ConnectorState isAcceptable(Widget widget, Point point, Transferable transferable) {
        Image dragImage = scene.getImageFromTransferable(transferable);
        JComponent view = scene.getView();
        Graphics2D g2 = (Graphics2D) view.getGraphics();
        Rectangle visRect = view.getVisibleRect();
        view.paintImmediately(visRect.x, visRect.y, visRect.width, visRect.height);
        g2.drawImage(dragImage,
                AffineTransform.getTranslateInstance(point.getLocation().getX(),
                point.getLocation().getY()),
                null);
        return ConnectorState.ACCEPT;
    }
}

File: platform/core/org.openmole.ide.core.implementation/src/main/scala/org/openmole/ide/core/implementation/provider/DnDProvider.scala
/*
 * Copyright (C) 2011 Mathieu leclaire <mathieu.leclaire at openmole.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.ide.core.implementation.provider

import java.awt.Point
import java.awt.datatransfer.Transferable
import java.awt.geom.AffineTransform
import org.netbeans.api.visual.action.AcceptProvider
import org.netbeans.api.visual.widget.Widget
import org.openmole.ide.core.implementation.workflow.Images
import org.openmole.ide.core.implementation.workflow.MoleScene
import org.netbeans.api.visual.action.ConnectorState
import java.awt.Graphics
import java.awt.Graphics2D
import org.openmole.ide.core.model.workflow.IMoleScene

abstract class DnDProvider(molescene: IMoleScene) extends AcceptProvider{

  override def isAcceptable(widget: Widget, point: Point,transferable:Transferable): ConnectorState= {
    val view= molescene.graphScene.getView
    val visRect = view.getVisibleRect
    val g2= view.getGraphics.asInstanceOf[Graphics2D]
    view.paintImmediately(visRect.x, visRect.y, visRect.width, visRect.height)
    
    g2.drawImage(Images.getImageFromTransferable(transferable),
                               AffineTransform.getTranslateInstance(point.getLocation().getX(),point.getLocation().getY()),
                               null)
    return ConnectorState.ACCEPT
  }
}